### PR TITLE
Discord Login Controller

### PR DESCRIPTION
**Pull request by jacksonargo**
This is a shard from #94 

This adds a discord client with 3 methods
 * ~~QueryOAuth - returns a user token with identity scope using discord oauth api. This uses a discord oauth client id/secret.~~
 * QueryUser - Returns the user id of the bearer token. This uses a token returned by discord's oauth api.
 * QueryGuildRoles - Returns the guild roles of a user for a given user id and guild id. This uses the vvgo bot token.

### CODE REVIEW

**Comment by jacksonargo** [comment]
We get the authorization code when discord redirects the user to our server. That's handled by the DiscordLoginHandler

https://github.com/virtual-vgo/vvgo/pull/94/files#diff-156f70b66d0338bcbce16f6d606a738dR73

File: pkg/discord/discord.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+type UserID string
+
+func (x UserID) String() string { return string(x) }
+
+// A discord guild id.
+// These are string encoded 64bit ints.
+// https://discordapp.com/developers/docs/reference#snowflakes
+type GuildID string
+
+func (x GuildID) String() string { return string(x) }
```
[the remark this answers is not in the document]
Comment (jacksonargo) [reply pkg/discord/discord.go:75]: Discord sends us the ids as strings, and we send the back to discord as strings, so no reason to convert them to int64.

[the remark this answers is not in the document]
Comment (0az) [reply pkg/discord/discord.go:75]: Fair enough.

**Comment by jacksonargo** [comment]
> we need to check the callback's returned state against an expected value

Can you please elaborate on this? How does this work?

**Comment by jacksonargo** [comment]
Ah I see, you are talking about this: https://discordapp.com/developers/docs/topics/oauth2#state-and-security ?

**Comment by jacksonargo** [comment]
Ah ok I think I get how this works.
1. We advertise a link like https://vvgo.org/login_with_discord
2. The user clicks that link, and which is handled by, say, api.DiscordPreAuth.
3. DiscordPreAuth creates a secret status and adds it as a query param to the real discord login link, like:

```
https://discordapp.com/api/oauth2/authorize?status=94c710cba753c8c25289&client_id=700963768787795998&redirect_uri=https%3A%2F%2Fjackson.vvgo.org%2Fauth%2Fdiscord&response_type=code&scope=identify
```

4. The user clicks that link, authorizes with discord, and then discord redirects them back to our server via this link, which includes our status:
```
https://vvgo.org/auth/discord?state=94c710cba753c8c25289&code=80351110224678912
```

5. The api server's DiscordLoginHandler receives this request and verifies that the request came from the same browser session that originally hit the DiscordPreAuth handler.

```

Does that sound right?

**Comment by jacksonargo** [comment]
Updated to use cookie when tracking request states.

**Comment by jacksonargo** [comment]
Implementing the cookie -> state bit, I think can be handled entirely with the cookie data.
Our cookies are cryptographically tamper safe with expiration times.
* The secrets will be a cryptorand int64.
* We can store these oauth session secrets in a broweser cookie named `vvgo-oauth-state` with a 5 min expiration.

**Comment by jacksonargo** [comment]
Actually, we'll use your algorithm -> https://github.com/virtual-vgo/vvgo/issues/101#issuecomment-616761600

I don't know how to do it with jwt's, but I can implement with cookies.

**Comment by jacksonargo** [comment]
Darn I really need to write the unit tests for this controller too.

File: pkg/api/login.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			return false
+		}
+		return true
+	}
+
+	// read the oauth token from the request
+	var oauthToken discord.OAuthToken
+	if ok := handleError(json.NewDecoder(r.Body).Decode(&oauthToken)); !ok {
+		return
+	}
```
Comment (0az) [inline pkg/api/login.go:46]: Honestly, I'd rather just store the token in Redis, or just have this take a Discord User Snowflake from Redis. Either way, it'd be attached to the user's pre-session map.

All of this gets tricky once we start worrying about this plus Username/Pass authentication, though.

**Comment by 0az** [comment]
RESOLVED:
- Golang API has responsibility for **Authorization**
- Python API has responsibility for **Authentication**

**Comment by jacksonargo** [comment]
Implemented the stateful redirect

**Pull request merged**